### ISSUE TITLE
Need to compensate for Azure Pipelines' changed default to use shallow clones

### ISSUE THREAD

**Issue by SteveBush** [AArnott/Library.Template#176]
I created a new template and configured GitHub Actions and Azure Pipelines to run it.   I receive the following error in the Set build number step.  All of my other pipelines work fine.  The pipeline works fine on Github Actions but fails on Azure Pipelines with the following error:

```
2022-09-07T07:00:09.0723419Z ##[section]Starting: ⚙ Set build number
2022-09-07T07:00:09.0933050Z ==============================================================================
2022-09-07T07:00:09.0933261Z Task         : PowerShell
2022-09-07T07:00:09.0933434Z Description  : Run a PowerShell script on Linux, macOS, or Windows
2022-09-07T07:00:09.0933588Z Version      : 2.200.0
2022-09-07T07:00:09.0933719Z Author       : Microsoft Corporation
2022-09-07T07:00:09.0933914Z Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/powershell
2022-09-07T07:00:09.0934151Z ==============================================================================
2022-09-07T07:00:10.1678360Z Generating script.
2022-09-07T07:00:10.2024735Z ========================== Starting Command Output ===========================
2022-09-07T07:00:10.2195711Z ##[command]"C:\windows\System32\WindowsPowerShell\v1.0\powershell.exe" -NoLogo -NoProfile -NonInteractive -ExecutionPolicy Unrestricted -Command ". 'C:\AzureAgent\_work\_temp\da3636f8-5d4c-4cd7-9eb5-0db983b2fd54.ps1'"
2022-09-07T07:00:10.7001249Z Installing nbgv to C:\AzureAgent\_work\_temp\1508
2022-09-07T07:00:15.1309423Z Unhandled exception: System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation.
2022-09-07T07:00:15.1310763Z  ---> Nerdbank.GitVersioning.GitException: Shallow clone lacks the objects required to calculate version height. Use full clones or clones with a history at least as deep as the last version height resetting change.
2022-09-07T07:00:15.1311980Z  ---> Nerdbank.GitVersioning.GitException: An commit object with SHA 5dd190be16d92bfe476878d32ecdce49e69edfc7 could not be found.
2022-09-07T07:00:15.1313185Z    at Nerdbank.GitVersioning.ManagedGit.GitRepository.GetObjectBySha(GitObjectId sha, String objectType) in D:\a\1\s\src\NerdBank.GitVersioning\ManagedGit\GitRepository.cs:line 505
2022-09-07T07:00:15.1314510Z    at Nerdbank.GitVersioning.ManagedGit.GitRepository.GetCommit(GitObjectId sha, Boolean readAuthor) in D:\a\1\s\src\NerdBank.GitVersioning\ManagedGit\GitRepository.cs:line 264
2022-09-07T07:00:15.1316527Z    at Nerdbank.GitVersioning.Managed.GitExtensions.<GetCommitHeight>g__TryCalculateHeight|4_0(GitCommit commit, <>c__DisplayClass4_0& ) in D:\a\1\s\src\NerdBank.GitVersioning\Managed\ManagedGitExtensions.cs:line 133
2022-09-07T07:00:15.1317411Z    at Nerdbank.GitVersioning.Managed.GitExtensions.GetCommitHeight(GitRepository repository, GitCommit startingCommit, GitWalkTracker tracker, Func`2 continueStepping) in D:\a\1\s\src\NerdBank.GitVersioning\Managed\ManagedGitExtensions.cs:line 207
2022-09-07T07:00:15.1318278Z    at Nerdbank.GitVersioning.Managed.GitExtensions.GetHeight(ManagedGitContext context, Func`2 continueStepping) in D:\a\1\s\src\NerdBank.GitVersioning\Managed\ManagedGitExtensions.cs:line 101
2022-09-07T07:00:15.1319556Z    at Nerdbank.GitVersioning.Managed.GitExtensions.GetVersionHeight(ManagedGitContext context, Version baseVersion) in D:\a\1\s\src\NerdBank.GitVersioning\Managed\ManagedGitExtensions.cs:line 54
2022-09-07T07:00:15.1320547Z    at Nerdbank.GitVersioning.Managed.ManagedGitContext.CalculateVersionHeight(VersionOptions committedVersion, VersionOptions workingVersion) in D:\a\1\s\src\NerdBank.GitVersioning\Managed\ManagedGitContext.cs:line 105
2022-09-07T07:00:15.1321553Z    at Nerdbank.GitVersioning.VersionOracle..ctor(GitContext context, ICloudBuild cloudBuild, Nullable`1 overrideVersionHeightOffset) in D:\a\1\s\src\NerdBank.GitVersioning\VersionOracle.cs:line 66
2022-09-07T07:00:15.1322181Z    --- End of inner exception stack trace ---
2022-09-07T07:00:15.1322776Z    at Nerdbank.GitVersioning.VersionOracle.<.ctor>g__ThrowShallowClone|5_0(Exception inner) in D:\a\1\s\src\NerdBank.GitVersioning\VersionOracle.cs:line 79
2022-09-07T07:00:15.1323738Z    at Nerdbank.GitVersioning.VersionOracle..ctor(GitContext context, ICloudBuild cloudBuild, Nullable`1 overrideVersionHeightOffset) in D:\a\1\s\src\NerdBank.GitVersioning\VersionOracle.cs:line 71
2022-09-07T07:00:15.1324954Z    at Nerdbank.GitVersioning.Commands.CloudCommand.SetBuildVariables(String projectDirectory, IEnumerable`1 metadata, String version, String ciSystem, Boolean allVars, Boolean commonVars, IEnumerable`1 additionalVariables, Boolean alwaysUseLibGit2) in D:\a\1\s\src\NerdBank.GitVersioning\Commands\CloudCommand.cs:line 116
2022-09-07T07:00:15.1328185Z    at Nerdbank.GitVersioning.Tool.Program.OnCloudCommand(String project, IReadOnlyList`1 metadata, String version, String ciSystem, Boolean allVars, Boolean commonVars, IReadOnlyList`1 define) in D:\a\1\s\src\nbgv\Program.cs:line 604
2022-09-07T07:00:15.1328899Z    --- End of inner exception stack trace ---
2022-09-07T07:00:15.1329394Z    at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
2022-09-07T07:00:15.1330081Z    at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
2022-09-07T07:00:15.1330577Z    at System.Delegate.DynamicInvokeImpl(Object[] args)
2022-09-07T07:00:15.1330961Z    at System.Delegate.DynamicInvoke(Object[] args)
2022-09-07T07:00:15.1331475Z    at System.CommandLine.Invocation.ModelBindingCommandHandler.InvokeAsync(InvocationContext context)
2022-09-07T07:00:15.1332092Z    at System.CommandLine.Invocation.InvocationPipeline.<>c__DisplayClass4_0.<<BuildInvocationChain>b__0>d.MoveNext()
2022-09-07T07:00:15.1332664Z --- End of stack trace from previous location where exception was thrown ---
2022-09-07T07:00:15.1333244Z    at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseParseErrorReporting>b__21_0>d.MoveNext()
2022-09-07T07:00:15.1333823Z --- End of stack trace from previous location where exception was thrown ---
2022-09-07T07:00:15.1334388Z    at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass16_0.<<UseHelp>b__0>d.MoveNext()
2022-09-07T07:00:15.1334956Z --- End of stack trace from previous location where exception was thrown ---
2022-09-07T07:00:15.1335556Z    at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass25_0.<<UseVersionOption>b__0>d.MoveNext()
2022-09-07T07:00:15.1336147Z --- End of stack trace from previous location where exception was thrown ---
2022-09-07T07:00:15.1336739Z    at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass23_0.<<UseTypoCorrections>b__0>d.MoveNext()
2022-09-07T07:00:15.1337320Z --- End of stack trace from previous location where exception was thrown ---
2022-09-07T07:00:15.1337890Z    at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseSuggestDirective>b__22_0>d.MoveNext()
2022-09-07T07:00:15.1338778Z --- End of stack trace from previous location where exception was thrown ---
2022-09-07T07:00:15.1339384Z    at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseParseDirective>b__20_0>d.MoveNext()
2022-09-07T07:00:15.1339937Z --- End of stack trace from previous location where exception was thrown ---
2022-09-07T07:00:15.1340407Z    at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseDebugDirective>b__11_0>d.MoveNext()
2022-09-07T07:00:15.1340884Z --- End of stack trace from previous location where exception was thrown ---
2022-09-07T07:00:15.1341387Z    at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<RegisterWithDotnetSuggest>b__10_0>d.MoveNext()
2022-09-07T07:00:15.1341909Z --- End of stack trace from previous location where exception was thrown ---
2022-09-07T07:00:15.1343245Z    at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass14_0.<<UseExceptionHandler>b__0>d.MoveNext()
2022-09-07T07:00:15.1977485Z ##[section]Finishing: ⚙ Set build number
```

I'm struggling to figure out why this error doesn't happen locally or on GitHub Actions, just Azure Pipelines.

Any pointers or help would be appreciated.

**Comment by SteveBush** [comment]
Woke up this morning and figured out the problem is the checkout fetch depth of the GitHub repository on Azure Pipelines.  Added fetchDepth: 0 to checkout step in azure-pipelines\build.yml and now everything works.  

```
jobs:
- job: Windows
  timeoutInMinutes: ${{ parameters.jobTimeout }}
  pool: ${{ parameters.windowsPool }}
  variables:
    BuildConfiguration: Release-Windows
  steps:
  - checkout: self
    clean: true
    fetchDepth: 0
  - template: install-dependencies.yml
    parameters:
      initArgs: -NoRestore
      installWorkLoads: maui, macos, android

  - powershell: '& (./azure-pipelines/Get-nbgv.ps1) cloud -c'
    displayName:  ⚙ Set build number

  - template: dotnet.yml
    parameters:
      RunTests: ${{ parameters.RunTests }}
```

**Comment by AArnott** [comment]
Excellent. Then this is by design, and the error message indicated the problem. :)

**Comment by AArnott** [comment]
For my own education though, I'm curious about `fetchDepth: 0`. I've never had to specify that in Azure Pipelines before, even for my github repos. Did you do something elsewhere in your Azure DevOps account to set a default fetch depth? I thought AzP would clone the full repo every time unless you explicitly set fetchDepth to some positive number.

**Comment by SteveBush** [comment]
I found a related issue in Azure checkout schemas

https://github.com/MicrosoftDocs/azure-devops-yaml-schema/issues/32#issue-1362216941

**Comment by AArnott** [comment]
Thanks for the link. There's no way that the default fetch depth is 1 for AzP, at least not in any of the accounts that I work with, because that would break _all_ my repositories. I've subscribed to that AzP issue to see what becomes of this.

**Comment by AArnott** [comment]
I see the default has in fact changed for *new* pipelines (even those that point to 'old' yaml files). I'll update the yaml file to enforce non-shallow clones.

**Comment by AArnott** [comment]
Fixed in 5d14d2cecbb3fd3caa6a421da1525d8480baef8b

**Comment by AArnott** [comment]
Also documented at Nerdbank.GitVersioning with https://github.com/dotnet/Nerdbank.GitVersioning/commit/efc3b4c6f005530d0576d79f16ae93b3bb62d512
